src/realm/object-store/sync/app.cpp

```diff
@@ -723,85 +744,180 @@ std::string App::url_for_path(const std::string& path = "") const
+                }
+            });
+        return; // early return
     }
 
-    // if we do not have metadata yet, we need to initialize it and send the
-    // request once that's complete
-    init_app_metadata([completion = std::move(completion), request = std::move(request),
-                       anchor = shared_from_this()](const util::Optional<Response>& error) mutable {
-        if (error) {
-            return completion(std::move(*error));
-        }
+    // if we do not have metadata yet, update the metadata and resend the request
+    update_metadata_and_resend(std::move(request), std::move(completion));
+}
 
-        // if this is the first time we have received app metadata, the
-        // original request will not have the correct URL hostname for
-        // non global deployments.
-        auto app_metadata = anchor->m_sync_manager->app_metadata();
-        if (app_metadata && app_metadata->deployment_model != "GLOBAL" &&
-            request.url.rfind(anchor->m_base_url, 0) != std::string::npos) {
-            request.url.replace(0, anchor->m_base_url.size(), app_metadata->hostname);
+void App::handle_redirect_response(Request&& request, const Response& response,
+                                   UniqueFunction<void(const Response&)>&& completion)
+{
+    // Permanent redirect - get the location and init the metadata again
+    // Look for case insensitive redirect "location" in headers
+    auto location = response.headers.find("location");
+    if (location == response.headers.end() || location->second.empty()) {
+        // Location not found in the response, pass error response up the chain
+        Response error;
+        error.http_status_code = response.http_status_code;
+        error.client_error_code = ClientErrorCode::redirect_error;
+        error.body = "Redirect response missing location header";
+        return completion(std::move(error)); // early return
+    }
+
+    if (request.max_redirects) {
+        // Make sure we don't do too many redirects (max_http_redirects (20) is an arbitrary number)
+        if (*(request.max_redirects) > max_http_redirects) {
+            Response error;
+            error.http_status_code = response.http_status_code;
+            error.custom_status_code = 0;
+            error.client_error_code = ClientErrorCode::too_many_redirects;
+            error.body = util::format("number of redirections exceeded %1", max_http_redirects);
+            return completion(std::move(error)); // early return
         }
+        *(request.max_redirects) += 1;
+    }
+    else {
+        request.max_redirects = 1;
+    }
 
-        anchor->m_config.transport->send_request_to_server(std::move(request), std::move(completion));
-    });
+    // Update the metadata using the new location and replay the request with the updated hostname
+    // Trim off any trailing path/anchor/query string
+    constexpr size_t https_len = std::char_traits<char>::length("https://");
```

Comment:
do we want to make sure the URL starts with `https://` before we use that? Maybe do:
```
std::string_view new_url = location->second;
if (auto scheme_end = new_url.find("://"); scheme_end != std::string_view::npos) {
    new_url.remove_prefix(scheme_end + std::char_traits<char>::length("://"));
}
```

Then  new_url would be the url minus the initial scheme.

Reply:
I wanted to skip past the scheme and look for the first `/`, `#`, or `?` after the hostname and trim just that part until the end.